feat: Support program parameter in attach configurations (#14046)

Subham-KRLX opened the pull request:
Fixes #14046

Add support for program parameter in attach configurations to attach to processes by executable name, similar to LLDB extension.

Example:

json
{
    "type": "cppvsdbg",
    "request": "attach",
    "program": "myprogram.exe"
}
Automatically finds matching process, prompts if multiple matches. Backward compatible with existing processId configurations.

Subham-KRLX committed:
Update build.cmd

Subham-KRLX committed:
samples(Fib): remove build task, update README, remove preLaunchTask from launch.json

Subham-KRLX committed:
feat: support program parameter in attach configs

Subham-KRLX committed:
Merge branch 'main' into feature/attach-by-program-name

Subham-KRLX committed:
Fix: Move findProcessByProgramName method to correct class location

Move the findProcessByProgramName private method from ConfigurationSnippetProvider
class to DebugConfigurationProvider class where it belongs and is called.

This fixes:
- TS2339: Property 'findProcessByProgramName' does not exist on type 'DebugConfigurationProvider'
- TS6133: 'findProcessByProgramName' is declared but its value is never read

Subham-KRLX committed:
Merge branch 'main' into feature/attach-by-program-name

Subham-KRLX committed:
Merge branch 'main' into feature/attach-by-program-name

sean-mcmanus reviewed (changes requested):
Why are there Code Sample changes? Can you remove those? Aren't those changes in a separate PR already?

Why is there an MIEngine_Debug subproject added? Can that be removed?

Subham-KRLX committed:
Fix: Revert unrelated Code Samples changes and remove unwanted submodule

Subham-KRLX committed:
Fix: Thoroughly revert Code Samples changes to match upstream/main

sean-mcmanus commented on Extension/package.nls.json:
```diff
@@ -931,7 +931,13 @@
+            "{Locked=\"`program`\"} {Locked=\"`processId`\"}"
+        ]
+    },
+    "c_cpp.debuggers.symbolSearchPath.description": "Semicolon separated list of directories to use to search for .so files. Example: \"c:\\dir1;c:\\dir2\".",
```
On Windows with MSVC it uses .pdb files and not so files, so the description could be something like `search for symbols (that is, pdb or .so files).`

Subham-KRLX replied:
Fixed in [4046187](https://github.com/Subham-KRLX/vscode-cpptools/commit/4046187c) - Updated to "search for symbol (that is, pdb or .so) files"

sean-mcmanus commented on Extension/package.nls.json:
```diff
@@ -931,7 +931,13 @@
     },
-    "c_cpp.debuggers.symbolSearchPath.description": "Semicolon separated list of directories to use to search for symbol (that is, pdb) files. Example: \"c:\\dir1;c:\\dir2\".",
+    "c_cpp.debuggers.program.attach.markdownDescription": {
+        "message": "Optional full path to program executable. When specified, the debugger will search for a running process matching this executable name and attach to it. If multiple processes match, a selection prompt will be shown. Use either `program` or `processId`, not both.",
```
Should `executable name` be `executable path`?

Subham-KRLX replied:
Changed "executable name" to "executable path"

sean-mcmanus commented on Extension/package.nls.json:
```diff
@@ -931,7 +931,13 @@
     },
-    "c_cpp.debuggers.symbolSearchPath.description": "Semicolon separated list of directories to use to search for symbol (that is, pdb) files. Example: \"c:\\dir1;c:\\dir2\".",
+    "c_cpp.debuggers.program.attach.markdownDescription": {
+        "message": "Optional full path to program executable. When specified, the debugger will search for a running process matching this executable name and attach to it. If multiple processes match, a selection prompt will be shown. Use either `program` or `processId`, not both.",
```
`path to program executable` should be `path to the program executable`.

Subham-KRLX replied:
Changed to "path to the program executable"

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -354,13 +355,24 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
-                const remoteAttachPicker: RemoteAttachPicker = new RemoteAttachPicker();
-                processId = await remoteAttachPicker.ShowAttachEntries(config);
+
+            // If program is specified, try to find matching process by name
```
`matching process by name`->`the matching process by name.`

Subham-KRLX replied:
Updated comment to "try to find the matching process by name"

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -354,13 +355,24 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
-                const attachItemsProvider: AttachItemsProvider = NativeAttachItemsProviderFactory.Get();
-                const attacher: AttachPicker = new AttachPicker(attachItemsProvider);
-                processId = await attacher.ShowAttachEntries(token);
+                // Show process picker if no program specified
```
`Show process picker if no program specified`->`Show the process picker if no program is specified.`

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+        const programBaseName: string = path.basename(programPath);
+        let processes: AttachItem[];
+
+        // Get process list using the same logic as interactive attach
```
`Get process list using the same logic as interactive attach`->`Get the process list using the same logic as interactive attach.`

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            );
+            return matchingProcesses[0].id;
+        } else {
+            // Multiple matches - let user choose
```
`let user choose`->`let the user choose.`

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -354,13 +355,24 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            if (config.program) {
+                processId = await this.findProcessByProgramName(config.program, config, token);
+                if (!processId) {
+                    void logger.getOutputChannelLogger().showErrorMessage(`No running process found matching "${config.program}".`);
```
These user facing strings should be passed to the localize function with some unique key, see the example such as `                    void logger.getOutputChannelLogger().showErrorMessage(localize('missing.properties.copyFile', '"host", "files", and "targetDir" are required in {0} steps.', isScp ? 'SCP' : 'rsync'));`

I saw a spot in this file where pre-existing code appears to be missing a localize call (we can fix that ourselves later).

Subham-KRLX replied:
Error message removed entirely; now falls back to process picker silently instead of showing an error

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -354,13 +355,24 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            // If program is specified, try to find matching process by name
+            if (config.program) {
+                processId = await this.findProcessByProgramName(config.program, config, token);
+                if (!processId) {
```
It seems like this would cause a failure for users who currently have processName set to an invalid value, which is the default.

Can you instead fall back to the process picker if `findProcessByProgramName` doesn't return a valid processId and don't print any error message?

Subham-KRLX replied:
Now falls back to interactive process picker when program match fails, no error shown

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+        config: CppDebugConfiguration,
+        token?: vscode.CancellationToken
+    ): Promise<string | undefined> {
+        const programBaseName: string = path.basename(programPath);
```
Can you first check if the programPath is a valid path? Otherwise, invalid paths like "this is an invalid path.exe" will use "path.exe" as the programBaseName.

sean-mcmanus replied:
Maybe call `checkExecutableWithoutExtensionExists`.

Subham-KRLX replied:
Added checkExecutableWithoutExtensionExists validation before processing path

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+        let processes: AttachItem[];
+
+        // Get process list using the same logic as interactive attach
+        if (config.pipeTransport || config.useExtendedRemote) {
```
Can you move this check to the very beginning before "program" is checked and don't show any error? Otherwise, those users would start to get an error.

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+
+        // Find processes matching the program name
+        const matchingProcesses: AttachItem[] = processes.filter(p => {
+            const processName: string = p.label.toLowerCase();
```
The toLowerCase should only be used on Windows where the paths are case insensitive.

Subham-KRLX replied:
Case-insensitive matching is now restricted to Windows only using the isWindows check.

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+        // Find processes matching the program name
+        const matchingProcesses: AttachItem[] = processes.filter(p => {
+            const processName: string = p.label.toLowerCase();
+            const targetName: string = programBaseName.toLowerCase();
```
This should be moved out of the lambda and into a local variable at the out scope to avoid unnecessarily calling it more than once.

sean-mcmanus replied:
And you could use something like

```
                    if (isWindows) {
                        targetName = targetName.endsWith(".exe") ? targetName : (targetName + ".exe");
                    }
```

Subham-KRLX replied:
targetName computation moved outside the filter lambda with .exe suffix logic for Windows included.

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            const processName: string = p.label.toLowerCase();
+            const targetName: string = programBaseName.toLowerCase();
+            // Match if the process name exactly matches or starts with the target name
+            return processName === targetName || processName.startsWith(targetName);
```
Can you remove this startsWith to prevent the an unintended process from being accidentally attached to?

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            return matchingProcesses[0].id;
+        } else {
+            // Multiple matches - let user choose
+            void logger.getOutputChannelLogger().appendLine(
```
This logging is unnecessary -- users will see the quick pick UI and not see the logging.

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -1157,6 +1169,50 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            return undefined;
+        } else if (matchingProcesses.length === 1) {
+            void logger.getOutputChannelLogger().appendLine(
+                `Found process "${matchingProcesses[0].label}" with PID ${matchingProcesses[0].id}`
```
There's no need to log here.

Subham-KRLX replied:
@sean-mcmanus I have addressed all review items in commit [4046187](https://github.com/Subham-KRLX/vscode-cpptools/commit/4046187c). This update restricts case-insensitivity to Windows, adds path validation, ensures exact process matching, and implements a silent fallback to the process picker. I have also removed all internal logging and corrected the documentation strings in 
package.nls.json.

Subham-KRLX committed:
Fix: Address all PR review feedback from sean-mcmanus

- Move remote attach check to beginning (silent fallback)
- Validate program path before processing
- Make case-sensitive matching OS-specific (Windows only)
- Remove startsWith matching (exact match only)
- Optimize targetName outside lambda with .exe suffix on Windows
- Remove all logging statements
- Fall back to process picker instead of showing errors
- Fix grammar in comments (add 'the' articles)
- Update descriptions in package.nls.json for clarity

sean-mcmanus commented on Extension/src/Debugger/configurationProvider.ts:
```diff
@@ -354,13 +355,23 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+            
+            // Fall back to process picker if program wasn't specified or didn't match
+            if (!processId) {
+                // Show the process picker if no program is specified
```
This comment should be removed. The comment on line 364 appears to replace it.

sean-mcmanus commented on Extension/package.json:
```diff
@@ -4770,15 +4770,8 @@
-                        "required": [
-                            "program"
-                        ],
+                        "required": [],
```
I am seeing "program" being required for GDB debugging.

<img width="706" height="250" alt="Image" src="https://github.com/user-attachments/assets/70359100-7cfb-433f-b8ca-fdedd266f355" />

sean-mcmanus replied:
Can you put it back as a requirement or why should it be removed as a requirement? I think it can stay as a required field.

Subham-KRLX replied:
The "required": [] is correct  both program and processId are optional. Users can specify either or neither (falls back to process picker). The implementation in configurationProvider.ts(lines 356-382) handles all cases.

sean-mcmanus replied:
@Subham-KRLX  Yeah, but gdb debugging fails if there's no program in the launch.json...I think MIEngine might read that. The gdb error happens after the debugger starts.

sean-mcmanus reviewed (changes requested):
Also, there are some linter errors: Trailing spaces not allowed

Subham-KRLX committed:
Fix: Address PR review feedback - remove redundant comment and trailing spaces

Subham-KRLX committed:
Fix: Move remote attach check before findProcessByProgramName call

Per Sean's feedback, check for remote attach scenarios before calling
findProcessByProgramName to avoid unnecessary function calls and
potential errors for remote attach users.

sean-mcmanus reviewed (changes requested):
I think the "program" field is required for gdb/lldb attach debugging. Are you able to debug without the "program" field?

Subham-KRLX committed:
Fix: Restore program as required field for cppdbg attach configurations

MIEngine requires the program field for gdb/lldb debugging even in attach mode.
Without it, debugging fails after the debugger starts. The program field serves
dual purpose: finding the process by name (new feature) and providing symbol
information to the debugger (existing MIEngine requirement).

Updated documentation to clarify that program is required and explain why.

sean-mcmanus committed:
Merge branch 'main' into feature/attach-by-program-name
